_posts/2023-04-30-openai-supabase-summary.md

```diff
@@ -0,0 +1,175 @@
+
+### 📝 Using the GPT-4 API to Summarize Data
+
+The second step in the pipeline passes the latest data to [OpenAI GPT-4]() to summarize it.
+It also compares it to the last summary that was created, ensuring we have some historic
+comparison.
+
+```python
+@step(enable_cache=True)
+def gpt_4_summarizer(
+    params: SummarizerParams, documents: List[str], context: StepContext
+) -> str:
+    """Summarizes the data using GPT-4."""
+    openai_secret = Client().get_secret("openai")
+    last_step = (
+        get_pipeline(context.pipeline_name)
+        .runs[0]
+        .get_step("generate_summary")
+    )
+    if len(last_step.outputs) == 0:
+        last_analysis = "No previous analysis found."
+    else:
+        last_analysis = last_step.output.read()
+
+    openai.api_key = openai_secret.secret_values["api_key"]
+
+    response = openai.ChatCompletion.create(
+        model="gpt-4",
+        messages=[
+            {"role": "system", "content": params.system_content},
+            {
+                "role": "user",
+                "content": params.prompt_preamble
+                + "\n"
+                + "Here is an example output:"
+                + "\n"
+                + params.prompt_example
+                + "\n"
+                + "Here is the data: "
+                + str(documents)
+                + "\n"
+                + "And here is the previous analysis: "
+                + last_analysis,
+            },
+        ],
+        temperature=0,
+        max_tokens=256,
+        top_p=1.0,
+        frequency_penalty=0.0,
+        presence_penalty=0.0,
+    )
+    return response.choices[0].message.content
+```
+
+The `gpt_4_summarizer` step is a custom ZenML step that leverages GPT-4 to generate summaries of the input data. It takes `SummarizerParams` and a list of documents as input, along with the `StepContext` to access previous pipeline runs.
+
+The step first retrieves the secret API key for OpenAI and checks for any previous analysis. If a previous analysis is found, it is included in the input to GPT-4. This allows GPT-4 to consider the previous analysis while generating the new summary.
+
+The step then calls the GPT-4 API with the specified parameters and input data, including the documents and the previous analysis. The API returns a summary, which is then returned as the output of the step.
+
+This custom step can be adapted for other use cases by modifying the input parameters, the GPT-4 model, or the API call. For example, you could use this step to generate summaries for different types of data, use a different language model, or customize the API call to suit your specific needs. By customizing this step, you can harness the power of GPT-4 and Zen
```

Comment:
```suggestion
This custom step can be adapted for other use cases by modifying the input parameters, the GPT-4 model, or the API call. For example, you could use this step to generate summaries for different types of data, use a different language model, or customize the API call to suit your specific needs. By customizing this step, you can harness the power of GPT-4 and ZenML.
```

Given the cost difference, you'd probably be better off using `chat-3.5-turbo` instead. It's more than capable enough to handle this task.